apps/services/license-api/src/app/modules/license/guards/licenseTypeScope.guard.ts

```diff
@@ -0,0 +1,40 @@
+import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common'
+import { Observable } from 'rxjs'
+import { Reflector } from '@nestjs/core'
+import { SCOPES_KEY, getRequest } from '@island.is/auth-nest-tools'
+
+@Injectable()
+export class LicenseTypeScopesGuard implements CanActivate {
+  constructor(private readonly reflector: Reflector) {}
+
+  canActivate(
+    context: ExecutionContext,
+  ): boolean | Promise<boolean> | Observable<boolean> {
+    const scopes = this.reflector.getAllAndOverride<string[]>(SCOPES_KEY, [
+      context.getHandler(),
+      context.getClass(),
+    ])
+
+    const request = getRequest(context)
+
+    if (scopes && !this.hasScope(scopes, request.auth?.scope)) {
+      return false
+    }
+
+    const licenseType = this.getLicenseTypeFromUrl(request.url)
+    const scopeActions = this.getScopeActions(request.auth?.scope ?? [])
+
+    return scopeActions.includes(licenseType)
+  }
+
+  private hasScope(needScopes: string[], haveScopes: string[] = []): boolean {
+    return needScopes.some((scope) => haveScopes.includes(scope))
+  }
```

Comment:
I don't like the string handling in guard logic. Might break in unexpected ways in the future.

I'd suggest changing it to:

1. Use a map of license types and scopes. The guard is can import this directly instead of getting it dynamically using Reflector.
2. Get the license type from the request by doing `request.params.licenseType` (better param name than `licenseId` imo).
3. Return false if the `licenseType` param is empty or if the user does not have the scope for it, eg `request.auth?.scope?.includes(licenseTypeToScope[licenseType])`

Reply:
ah yeah, that seems much better